In [78]:
import pandas as pd
import numpy as np

df = pd.read_csv('../data/NYC_Taxi_Cleaned_Analysis_Ready.csv')

# Descriptive Analytics

In [79]:
# Initial exploration of the dataset
df.describe()

,VendorID,passenger_count,trip_distance,RatecodeID,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee,duration_min,speed_mph,pickup_hour
count,3.946698e+06,3.946698e+06,3.946698e+06,3.946698e+06,3.946698e+06,3.946698e+06,3.946698e+06,3.946698e+06,3.946698e+06,3.946698e+06,3.946698e+06,3.946698e+06,3.946698e+06,3.946698e+06,3.946698e+06,3.189311e+06,3.946698e+06,3.946698e+06,3.946698e+06,3.946698e+06
mean,1.787345e+00,1.233373e+00,3.441244e+00,2.710091e+00,1.620718e+02,1.611536e+02,9.466660e-01,1.936859e+01,1.307855e+00,4.963577e-01,3.028975e+00,5.057680e-01,9.830288e-01,2.813628e+01,1.861057e+00,1.493874e-01,5.504911e-01,1.697423e+01,1.102928e+01,1.420994e+01
std,4.194354e-01,6.595587e-01,4.121952e+00,1.264134e+01,6.562439e+01,7.000420e+01,6.359776e-01,1.576509e+01,1.842253e+00,4.729320e-02,3.787247e+00,1.986163e+00,1.288213e-01,1.988695e+01,1.090462e+00,4.889902e-01,3.314068e-01,1.339776e+01,6.144650e+00,5.989917e+00
min,1.000000e+00,0.000000e+00,1.100000e-01,1.000000e+00,1.000000e+00,1.000000e+00,0.000000e+00,1.000000e-02,-1.739000e+01,-7.810000e+00,0.000000e+00,0.000000e+00,0.000000e+00,1.010000e+00,0.000000e+00,0.000000e+00,-7.500000e-01,1.000000e+00,1.000000e+00,0.000000e+00
25%,2.000000e+00,1.000000e+00,1.100000e+00,1.000000e+00,1.220000e+02,1.070000e+02,1.000000e+00,9.300000e+00,0.000000e+00,5.000000e-01,0.000000e+00,0.000000e+00,1.000000e+00,1.656000e+01,0.000000e+00,0.000000e+00,0.000000e+00,8.266667e+00,7.082405e+00,1.000000e+01
50%,2.000000e+00,1.000000e+00,1.900000e+00,1.000000e+00,1.610000e+02,1.620000e+02,1.000000e+00,1.420000e+01,0.000000e+00,5.000000e-01,2.450000e+00,0.000000e+00,1.000000e+00,2.186000e+01,2.500000e+00,0.000000e+00,7.500000e-01,1.338333e+01,9.486166e+00,1.500000e+01
75%,2.000000e+00,1.000000e+00,3.700000e+00,1.000000e+00,2.330000e+02,2.330000e+02,1.000000e+00,2.260000e+01,2.500000e+00,5.000000e-01,4.070000e+00,0.000000e+00,1.000000e+00,3.085000e+01,2.500000e+00,0.000000e+00,7.500000e-01,2.116667e+01,1.303040e+01,1.900000e+01
max,6.000000e+00,9.000000e+00,9.906000e+01,9.900000e+01,2.630000e+02,2.630000e+02,4.000000e+00,1.334020e+03,1.500000e+01,2.214000e+01,5.500000e+02,1.158700e+02,1.000000e+00,1.345710e+03,2.500000e+00,1.750000e+00,1.500000e+00,1.799667e+02,8.000000e+01,2.300000e+01


In [80]:
# First 10 rows of the dataset
df.head(10)

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,...,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee,duration_min,speed_mph,pickup_hour,day_of_week,is_weekend
0,2,2025-01-18 20:53:30,2025-01-18 21:00:47,1,0.97,1,N,238,166,1,...,1.0,13.32,0.0,0.0,0.00,7.283333,7.990847,20,Saturday,True
1,1,2025-01-25 11:12:51,2025-01-25 11:17:57,1,0.60,1,N,50,48,2,...,1.0,10.55,2.5,0.0,0.75,5.100000,7.058824,11,Saturday,True
2,1,2025-01-21 15:09:31,2025-01-21 15:19:02,1,0.80,1,N,236,237,1,...,1.0,16.65,2.5,0.0,0.00,9.516667,5.043783,15,Tuesday,False
3,2,2025-01-11 22:25:45,2025-01-11 22:34:22,2,1.93,1,N,231,68,1,...,1.0,20.37,2.5,0.0,0.75,8.616667,13.439072,22,Saturday,True
4,2,2025-01-04 23:37:07,2025-01-04 23:45:58,1,4.44,1,N,137,88,0,...,1.0,35.12,0.0,NaN,0.00,8.850000,30.101695,23,Saturday,True
5,1,2025-01-29 00:31:22,2025-01-29 00:38:27,1,1.10,1,N,140,263,1,...,1.0,15.48,2.5,0.0,0.00,7.083333,9.317647,0,Wednesday,False
6,2,2025-01-30 19:19:28,2025-01-30 19:30:09,1,1.33,1,N,170,230,2,...,1.0,17.95,2.5,0.0,0.75,10.683333,7.469579,19,Thursday,False
7,2,2025-01-23 15:59:29,2025-01-23 16:10:38,1,1.25,1,N,43,43,1,...,1.0,16.15,2.5,0.0,0.75,11.150000,6.726457,15,Thursday,False
8,1,2025-01-23 11:27:08,2025-01-23 11:43:02,1,1.70,1,N,100,113,1,...,1.0,23.55,2.5,0.0,0.75,15.900000,6.415094,11,Thursday,False
9,1,2025-01-19 15:42:39,2025-01-19 15:53:35,1,1.20,1,N,163,141,1,...,1.0,16.15,2.5,0.0,0.75,10.933333,6.585366,15,Sunday,True


In [81]:
# Column data types and non-null counts
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 3946698 entries, 0 to 3946697
Data columns (total 25 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   VendorID               int64  
 1   tpep_pickup_datetime   str    
 2   tpep_dropoff_datetime  str    
 3   passenger_count        int64  
 4   trip_distance          float64
 5   RatecodeID             int64  
 6   store_and_fwd_flag     str    
 7   PULocationID           int64  
 8   DOLocationID           int64  
 9   payment_type           int64  
 10  fare_amount            float64
 11  extra                  float64
 12  mta_tax                float64
 13  tip_amount             float64
 14  tolls_amount           float64
 15  improvement_surcharge  float64
 16  total_amount           float64
 17  congestion_surcharge   float64
 18  Airport_fee            float64
 19  cbd_congestion_fee     float64
 20  duration_min           float64
 21  speed_mph              float64
 22  pickup_hour            int64 

In [82]:
# Number of missing values in each column
df.isnull().sum()

VendorID                      0
tpep_pickup_datetime          0
tpep_dropoff_datetime         0
passenger_count               0
trip_distance                 0
RatecodeID                    0
store_and_fwd_flag            0
PULocationID                  0
DOLocationID                  0
payment_type                  0
fare_amount                   0
extra                         0
mta_tax                       0
tip_amount                    0
tolls_amount                  0
improvement_surcharge         0
total_amount                  0
congestion_surcharge          0
Airport_fee              757387
cbd_congestion_fee            0
duration_min                  0
speed_mph                     0
pickup_hour                   0
day_of_week                   0
is_weekend                    0
dtype: int64

In [83]:
# Percentage of missing values in each column
df.isnull().mean() * 100

VendorID                  0.000000
tpep_pickup_datetime      0.000000
tpep_dropoff_datetime     0.000000
passenger_count           0.000000
trip_distance             0.000000
RatecodeID                0.000000
store_and_fwd_flag        0.000000
PULocationID              0.000000
DOLocationID              0.000000
payment_type              0.000000
fare_amount               0.000000
extra                     0.000000
mta_tax                   0.000000
tip_amount                0.000000
tolls_amount              0.000000
improvement_surcharge     0.000000
total_amount              0.000000
congestion_surcharge      0.000000
Airport_fee              19.190397
cbd_congestion_fee        0.000000
duration_min              0.000000
speed_mph                 0.000000
pickup_hour               0.000000
day_of_week               0.000000
is_weekend                0.000000
dtype: float64

In [84]:
# Check for duplicate rows in the dataset
df.duplicated().sum()

np.int64(0)

In [85]:
# Display the minimum values for each numeric column in the dataset
df.min(numeric_only=True)

VendorID                     1
passenger_count              0
trip_distance             0.11
RatecodeID                   1
PULocationID                 1
DOLocationID                 1
payment_type                 0
fare_amount               0.01
extra                   -17.39
mta_tax                  -7.81
tip_amount                 0.0
tolls_amount               0.0
improvement_surcharge      0.0
total_amount              1.01
congestion_surcharge       0.0
Airport_fee                0.0
cbd_congestion_fee       -0.75
duration_min               1.0
speed_mph                  1.0
pickup_hour                  0
is_weekend               False
dtype: object

In [86]:
# Display the maximum values for each numeric column in the dataset
df.max(numeric_only=True)

VendorID                          6
passenger_count                   9
trip_distance                 99.06
RatecodeID                       99
PULocationID                    263
DOLocationID                    263
payment_type                      4
fare_amount                 1334.02
extra                          15.0
mta_tax                       22.14
tip_amount                    550.0
tolls_amount                 115.87
improvement_surcharge           1.0
total_amount                1345.71
congestion_surcharge            2.5
Airport_fee                    1.75
cbd_congestion_fee              1.5
duration_min             179.966667
speed_mph                      80.0
pickup_hour                      23
is_weekend                     True
dtype: object

In [87]:
# Display the percentiles for each numeric column in the dataset
percentiles = [0.10, 0.25, 0.50, 0.75, 0.90, 0.99]

df_numeric = df.select_dtypes(include=['int', 'float'])

display(df_numeric.quantile(percentiles))

,VendorID,passenger_count,trip_distance,RatecodeID,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee,duration_min,speed_mph,pickup_hour
0.10,1.0,1.0,0.70,1.0,68.0,54.0,0.0,7.2,0.00,0.5,0.00,0.00,1.0,13.44,0.0,0.00,0.00,5.266667,5.296863,6.0
0.25,2.0,1.0,1.10,1.0,122.0,107.0,1.0,9.3,0.00,0.5,0.00,0.00,1.0,16.56,0.0,0.00,0.00,8.266667,7.082405,10.0
0.50,2.0,1.0,1.90,1.0,161.0,162.0,1.0,14.2,0.00,0.5,2.45,0.00,1.0,21.86,2.5,0.00,0.75,13.383333,9.486166,15.0
0.75,2.0,1.0,3.70,1.0,233.0,233.0,1.0,22.6,2.50,0.5,4.07,0.00,1.0,30.85,2.5,0.00,0.75,21.166667,13.030404,19.0
0.90,2.0,2.0,8.70,1.0,239.0,239.0,2.0,38.7,4.25,0.5,6.65,0.00,1.0,50.51,2.5,0.00,0.75,32.300000,18.941860,21.0
0.99,2.0,4.0,19.27,99.0,263.0,263.0,4.0,70.0,7.50,0.5,17.19,6.94,1.0,101.65,2.5,1.75,0.75,69.383333,33.356289,23.0


In [88]:
# Display the number of unique values for each column in the dataset
df.nunique()

VendorID                       3
tpep_pickup_datetime     3614373
tpep_dropoff_datetime    3612982
passenger_count               10
trip_distance               3521
RatecodeID                     7
store_and_fwd_flag             2
PULocationID                 258
DOLocationID                 257
payment_type                   5
fare_amount                 9185
extra                         65
mta_tax                       15
tip_amount                  3657
tolls_amount                 998
improvement_surcharge          3
total_amount               15388
congestion_surcharge           4
Airport_fee                    2
cbd_congestion_fee             6
duration_min                8077
speed_mph                 804048
pickup_hour                   24
day_of_week                    7
is_weekend                     2
dtype: int64

In [89]:
# Display the number of rows in the dataset
df.shape[0]

3946698

In [90]:
# Display the number of columns in the dataset
df.shape[1]

25

In [91]:
# Display the earliest and latest pickup dates in the dataset
pickup_dates = pd.to_datetime(df['tpep_pickup_datetime'])

print(f"Earliest Pickup: {pickup_dates.min()}")
print(f"Latest Pickup:   {pickup_dates.max()}")

Earliest Pickup: 2007-12-05 18:45:00
Latest Pickup:   2025-11-30 23:59:59


In [92]:
df['tpep_pickup_datetime'] = pd.to_datetime(df['tpep_pickup_datetime'])
df['pickup_year'] = df['tpep_pickup_datetime'].dt.year

print("Total non-null pickup_year values:", df['pickup_year'].notna().sum())
print("\nFull year breakdown:")
print(df['pickup_year'].value_counts().sort_index())

Total non-null pickup_year values: 3946698

Full year breakdown:
pickup_year
2007          1
2009          1
2024          1
2025    3946695
Name: count, dtype: int64


## Initial Observations & Notable Findings

### 1. Dataset Scale & Volume
* **High Transaction Volume:** The dataset contains approximately **3.94 million records** (3,946,698 trips), providing a dense sample size across all operating hours and zones.
* **Granularity:** Records capture individual trip events containing pickup/drop-off spatial identifiers, metered timestamps, fee breakdowns, and payment attributes.
* **Temporal Boundary Anomalies:** While the dataset targets 2025 operations (3,946,695 records), timestamp inspection identified 3 anomalous historical records dating back to 2007 (2007-12-05), 2009, and 2024 due to meter clock errors.

### 2. Missing Value Patterns & Structural Gaps
* **Selective Null Clustering:** Over **757,000 records** exhibit missing values in `Airport_fee`. This pattern corresponds structurally to standard street hails that do not originate at airport terminals (JFK or LaGuardia).
* **Unrecorded Operational Expenses:** The dataset does not track vehicle operating expenditures (fuel consumption, vehicle depreciation, maintenance, or vehicle lease fees).
* **Payment-Restricted Tipping:** Recorded values in `tip_amount` are populated exclusively for credit card transactions (`payment_type == 1`); cash tips are systematically unrecorded and register as zero.

### 3. Data Representation & Formatting Nuances
* **Nominal IDs as Numeric Types:** Identifier columns (such as `VendorID`, `RatecodeID`, and `PULocationID`) are stored as numeric/floating-point types rather than categorical variables.
* **Distorted Descriptive Statistics:** Standard summary methods calculate mathematically invalid metrics for identifiers (e.g., a mean `VendorID` of 1.79) and render counts in scientific notation (`3.946698e+06`).
* **Variable Interaction Sensitivity:** Mathematical comparisons and column-wise evaluations are sensitive to mixed column data types across boolean flags and numeric values.
* **Invalid Categorical Codes & Zero-Occupancy Logs:** Minimum and percentile checks revealed trips logged with 0 passengers (passenger_count) and unmapped rate codes (RatecodeID = 99 at the 99th percentile), which lie outside the standard NYC TLC 1–6 rate definitions.

### 4. Revenue Composition, Anomalies & Fee Friction
* **Negative Charge Artifacts:** Descriptive statistics (`.min()` and `.describe()`) revealed negative values in certain fee columns, specifically `extra` (-17.39), `mta_tax` (-7.81), and `cbd_congestion_fee` (-0.75). These anomalies logically represent meter dispute corrections, refunded fares, or system resets rather than standard operational trips.
* **Gross Revenue Inflation:** Raw fare totals (`total_amount`) reflect gross passenger charges rather than driver take-home earnings due to stacked regulatory surcharges (`cbd_congestion_fee`, `congestion_surcharge`, `mta_tax`, `improvement_surcharge`, and `Airport_fee`).
* **Fee Disproportionality:** Flat-rate fees represent a significantly higher proportion of short, low-fare trips compared to long-distance fares.
* **Congestion and Speed Variance:** Wide spreads in `duration_min` and `speed_mph` indicate substantial traffic friction across different pickup locations, suggesting that trip distance alone does not reliably predict trip duration or occupied earning efficiency.
* **Extreme Upper-Tail Spikes:** Maximum value checks exposed severe right-skewed outliers, including gross fares up to $1,345.71 (fare_amount of $1,334.02 and tip_amount of $550.00). When divided by short trip durations, these anomalies inflate net_revenue_per_hour to an unrealistic maximum of $29,025.00/hr; the ABT construction step below caps this at the 99th percentile prior to modeling.

# ABT Construction

In [93]:
# 1. Drop rows with anomalous pickup years
df['tpep_pickup_datetime'] = pd.to_datetime(df['tpep_pickup_datetime'])
df = df[df['tpep_pickup_datetime'].dt.year == 2025].copy()

# 2. Clean missing fees
df['Airport_fee'] = df['Airport_fee'].fillna(0)

# 3. Revenue & Fee Impact Metrics (For Q1, Q2, Q3)
# Neutralize negative fee anomalies (refund/dispute artifacts) before they enter total_fees
fee_cols = ['cbd_congestion_fee', 'congestion_surcharge', 'Airport_fee', 'mta_tax', 'improvement_surcharge']
df[fee_cols] = df[fee_cols].clip(lower=0)

# Calculate total fees (all non-driver regulatory pass-through charges) and the true net revenue
df['total_fees'] = df[fee_cols].sum(axis=1)
df['net_revenue'] = df['total_amount'] - df['total_fees']

# Calculate Net Revenue Per Occupied Driving Hour (guarded against duration_min == 0)
df['net_revenue_per_hour'] = np.where(
    df['duration_min'] > 0,
    (df['net_revenue'] / df['duration_min']) * 60,
    np.nan
)

# Cap extreme upper-tail outliers (e.g. the $29,025/hr spike) at the 99th percentile prior to modeling
cap_value = df['net_revenue_per_hour'].quantile(0.99)
df['net_revenue_per_hour'] = df['net_revenue_per_hour'].clip(upper=cap_value)

# Calculate the % of gross revenue eaten by fees (Extent of reduction for Q2)
df['fee_impact_pct'] = (df['total_fees'] / df['total_amount']) * 100

# 4. Engineer Fare Tiers & Trip Length Thresholds (For Q2)
# Using qcut to divide fares and distances into 3 equal-sized distribution buckets
df['fare_tier'] = pd.qcut(df['total_amount'], q=3, labels=['Low Fare', 'Medium Fare', 'High Fare'])
df['trip_length_tier'] = pd.qcut(df['trip_distance'], q=3, labels=['Short Trip', 'Medium Trip', 'Long Trip'])

# 5. Engineer Traffic Congestion Levels (For Q3)
# Defining traffic by speed limits/averages (0-8mph = Heavy, 8-15mph = Moderate, 15mph+ = Light)
# include_lowest=True captures trips at exactly 0 mph (pd.cut's default intervals are right-closed)
speed_bins = [0, 8, 15, 100]
speed_labels = ['Heavy Traffic', 'Moderate Traffic', 'Light Traffic']
df['congestion_level'] = pd.cut(df['speed_mph'], bins=speed_bins, labels=speed_labels, include_lowest=True)

# 6. Preview the variables needed for final analysis
cols_to_show = [
    'day_of_week', 'pickup_hour', 'PULocationID',
    'fare_tier', 'trip_length_tier', 'congestion_level',
    'fee_impact_pct', 'net_revenue_per_hour'
]

display(df[cols_to_show].head(10))

,day_of_week,pickup_hour,PULocationID,fare_tier,trip_length_tier,congestion_level,fee_impact_pct,net_revenue_per_hour
0,Saturday,20,238,Low Fare,Short Trip,Heavy Traffic,11.261261,97.372998
1,Saturday,11,50,Low Fare,Short Trip,Heavy Traffic,45.023697,68.235294
2,Tuesday,15,236,Low Fare,Short Trip,Heavy Traffic,24.024024,79.754816
3,Saturday,22,231,Medium Fare,Medium Trip,Moderate Traffic,23.318606,108.765957
4,Saturday,23,137,High Fare,Long Trip,Light Traffic,4.271071,223.783784
5,Wednesday,0,140,Low Fare,Short Trip,Moderate Traffic,25.839793,97.242353
6,Thursday,19,170,Low Fare,Short Trip,Heavy Traffic,26.462396,74.134165
7,Thursday,15,43,Low Fare,Short Trip,Heavy Traffic,29.411765,61.345291
8,Thursday,11,100,Medium Fare,Medium Trip,Heavy Traffic,20.169851,70.943396
9,Sunday,15,163,Low Fare,Short Trip,Heavy Traffic,29.411765,62.560976


## Analytical Base Table (ABT) Construction & Feature Engineering Justification

### 1. Dropping Rows with Anomalous Pickup Years
* **Enforcing Temporal Scope:** The study specifically evaluates fleet operations under the post-2025 congestion pricing regime, requiring all records to reflect 2025 traffic and fee structures.
* **Removing Meter Clock Errors:** Initial descriptive inspection revealed 3 corrupted records from 2007, 2009, and 2024 out of 3,946,698 total trips. Filtering strictly for `dt.year == 2025` eliminates these hardware timestamp errors, leaving 3,946,695 valid trips and preventing massive date gaps during chronological splitting and time-series lag generation.

### 2. Imputing Missing Airport Fees with Zero (`fillna(0)`)
* **Resolving Structural Nulls:** Initial inspection showed 757,387 missing values in the `Airport_fee` column. These nulls are not missing data errors; they represent standard non-airport street hails where the fee does not apply.
* **Preventing Arithmetic Null Propagation:** In Pandas, adding a number to a `NaN` value results in `NaN`. Filling these structural nulls with `0` ensures that the subsequent `total_fees` addition executes cleanly across all 3,946,695 rows without wiping out non-airport trip records.

### 3. Engineering Revenue & Fee Impact Metrics (Target & AQ1–AQ3)
* **Neutralizing Negative Fee Anomalies:** The negative values previously observed in `extra`, `mta_tax`, and `cbd_congestion_fee` (meter dispute corrections, refunds, or system resets) are clipped to a floor of `0` across all five fee columns before they enter `total_fees`, so refund artifacts can't artificially inflate the resulting `net_revenue`.
* **Isolating True Driver Take-Home Pay (`total_fees` & `net_revenue`):** Gross `total_amount` inflates apparent earnings because it includes all five mandatory regulatory pass-through surcharges identified earlier — `cbd_congestion_fee`, `congestion_surcharge`, `Airport_fee`, `mta_tax`, and `improvement_surcharge` — none of which belong to the driver. Subtracting the full `total_fees` sum from `total_amount` isolates the true net revenue retained from the trip.
* **Standardizing for Opportunity Cost (`net_revenue_per_hour`):** A high-fare trip is unprofitable if severe gridlock causes it to take an hour to complete. Dividing `net_revenue` by `duration_min` and scaling by 60 converts raw trip payouts into an hourly earning velocity ($/hr), establishing the primary target variable needed to optimize shift scheduling and zone positioning (AQ1). The division is guarded so that any `duration_min == 0` row returns `NaN` instead of an infinite value, and the resulting distribution is capped at its 99th percentile to remove the extreme right-tail spikes (e.g., the $29,025/hr outlier) flagged during descriptive analytics, preventing them from destabilizing model training.
* **Quantifying Fee Erosion (`fee_impact_pct`):** Dividing `total_fees` by `total_amount` (multiplied by 100) measures the exact percentage of gross revenue consumed by regulatory charges. Because congestion fees are flat charges, this metric is essential for proving how fees disproportionately penalize specific trips and zones (AQ2).

### 4. Engineering Fare Tiers & Trip Length Thresholds (`pd.qcut`)
* **Creating Balanced Analytical Buckets:** Both `total_amount` and `trip_distance` are heavily right-skewed continuous variables. Using quantile-based discretization (`pd.qcut` with `q=3`) partitions the dataset into three equal-sized tertiles (`Low Fare`, `Medium Fare`, `High Fare` and `Short Trip`, `Medium Trip`, `Long Trip`) rather than relying on arbitrary dollar or mileage cutoffs.
* **Supporting Threshold Rule Extraction (AQ2):** Grouping trips into standardized tiers allows direct evaluation of how the 2025 congestion fee erodes margins across different trip profiles, providing clear boundaries to establish minimum-fare and trip-length guidelines for drivers.

### 5. Engineering Traffic Congestion Levels (`pd.cut`)
* **Translating Speed into Operational States:** Raw `speed_mph` fluctuates continuously, making it difficult to formulate high-level dispatch rules. Binning speeds into domain-relevant thresholds—`0–8 mph` (`Heavy Traffic`), `8–15 mph` (`Moderate Traffic`), and `15–100 mph` (`Light Traffic`)—converts continuous velocity into interpretable urban traffic states. `include_lowest=True` ensures trips clocked at exactly `0 mph` fall into `Heavy Traffic` instead of being silently dropped as `NaN`, since `pd.cut`'s default intervals are right-closed (`(0, 8]`, not `[0, 8]`).
* **Identifying Gridlock Tipping Points (AQ3):** These discrete tiers allow the analysis to pinpoint the exact congestion state where pickup zones experience a sharp drop in `net_revenue_per_hour`, guiding the fleet on when to deprioritize gridlocked zones.

In [94]:
# Display the first 10 rows of the ABT
df.head(10)

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,...,day_of_week,is_weekend,pickup_year,total_fees,net_revenue,net_revenue_per_hour,fee_impact_pct,fare_tier,trip_length_tier,congestion_level
0,2,2025-01-18 20:53:30,2025-01-18 21:00:47,1,0.97,1,N,238,166,1,...,Saturday,True,2025,1.50,11.82,97.372998,11.261261,Low Fare,Short Trip,Heavy Traffic
1,1,2025-01-25 11:12:51,2025-01-25 11:17:57,1,0.60,1,N,50,48,2,...,Saturday,True,2025,4.75,5.80,68.235294,45.023697,Low Fare,Short Trip,Heavy Traffic
2,1,2025-01-21 15:09:31,2025-01-21 15:19:02,1,0.80,1,N,236,237,1,...,Tuesday,False,2025,4.00,12.65,79.754816,24.024024,Low Fare,Short Trip,Heavy Traffic
3,2,2025-01-11 22:25:45,2025-01-11 22:34:22,2,1.93,1,N,231,68,1,...,Saturday,True,2025,4.75,15.62,108.765957,23.318606,Medium Fare,Medium Trip,Moderate Traffic
4,2,2025-01-04 23:37:07,2025-01-04 23:45:58,1,4.44,1,N,137,88,0,...,Saturday,True,2025,1.50,33.62,223.783784,4.271071,High Fare,Long Trip,Light Traffic
5,1,2025-01-29 00:31:22,2025-01-29 00:38:27,1,1.10,1,N,140,263,1,...,Wednesday,False,2025,4.00,11.48,97.242353,25.839793,Low Fare,Short Trip,Moderate Traffic
6,2,2025-01-30 19:19:28,2025-01-30 19:30:09,1,1.33,1,N,170,230,2,...,Thursday,False,2025,4.75,13.20,74.134165,26.462396,Low Fare,Short Trip,Heavy Traffic
7,2,2025-01-23 15:59:29,2025-01-23 16:10:38,1,1.25,1,N,43,43,1,...,Thursday,False,2025,4.75,11.40,61.345291,29.411765,Low Fare,Short Trip,Heavy Traffic
8,1,2025-01-23 11:27:08,2025-01-23 11:43:02,1,1.70,1,N,100,113,1,...,Thursday,False,2025,4.75,18.80,70.943396,20.169851,Medium Fare,Medium Trip,Heavy Traffic
9,1,2025-01-19 15:42:39,2025-01-19 15:53:35,1,1.20,1,N,163,141,1,...,Sunday,True,2025,4.75,11.40,62.560976,29.411765,Low Fare,Short Trip,Heavy Traffic


In [95]:
df.to_csv('../data/NYC_Taxi_ABT.csv', index=False)